miniapps/meshing/tmop-fitting.cpp

```diff
@@ -0,0 +1,897 @@
+   else if (surf_ls_type == 2) // reactor
+   {
+      ls_coeff = new FunctionCoefficient(reactor);
+   }
+   else if (surf_ls_type == 6) // 3D shape
+   {
+      ls_coeff = new FunctionCoefficient(csg_cubecylsph);
+   }
+   else
+   {
+      MFEM_ABORT("Surface fitting level set type not implemented yet.")
+   }
+
+   ParMesh *pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
+   delete mesh;
+   for (int lev = 0; lev < rp_levels; lev++) { pmesh->UniformRefinement(); }
+
+   // 4. Setup background mesh for surface fitting
+   ParMesh *pmesh_surf_fit_bg = NULL;
+   if (surf_bg_mesh)
+   {
+      Mesh *mesh_surf_fit_bg = NULL;
+      if (dim == 2)
+      {
+         mesh_surf_fit_bg =
+            new Mesh(Mesh::MakeCartesian2D(4, 4, Element::QUADRILATERAL, true));
+      }
+      else if (dim == 3)
+      {
+         mesh_surf_fit_bg =
+            new Mesh(Mesh::MakeCartesian3D(4, 4, 4, Element::HEXAHEDRON, true));
+      }
+      mesh_surf_fit_bg->EnsureNCMesh();
+      pmesh_surf_fit_bg = new ParMesh(MPI_COMM_WORLD, *mesh_surf_fit_bg);
+      delete mesh_surf_fit_bg;
+   }
+
+   // 5. Define a finite element space on the mesh. Here we use vector finite
+   //    elements which are tensor products of quadratic finite elements. The
+   //    number of components in the vector finite element space is specified by
+   //    the last parameter of the FiniteElementSpace constructor.
+   FiniteElementCollection *fec;
+   if (mesh_poly_deg <= 0)
+   {
+      fec = new QuadraticPosFECollection;
+      mesh_poly_deg = 2;
+   }
+   else { fec = new H1_FECollection(mesh_poly_deg, dim); }
+   ParFiniteElementSpace *pfespace =
+      new ParFiniteElementSpace(pmesh, fec, dim, mesh_node_ordering);
+
+   // 6. Make the mesh curved based on the above finite element space. This
+   //    means that we define the mesh elements through a fespace-based
+   //    transformation of the reference element.
+   pmesh->SetNodalFESpace(pfespace);
+
+   // 7. Get the mesh nodes (vertices and other degrees of freedom in the finite
+   //    element space) as a finite element grid function in fespace. Note that
+   //    changing x automatically changes the shapes of the mesh elements.
+   ParGridFunction x(pfespace);
+   pmesh->SetNodalGridFunction(&x);
+   x.SetTrueVector();
+
+   // 10. Save the starting (prior to the optimization) mesh to a file. This
+   //     output can be viewed later using GLVis: "glvis -m perturbed -np
+   //     num_mpi_tasks".
+   {
+      ostringstream mesh_name;
+      mesh_name << "perturbed.mesh";
+      ofstream mesh_ofs(mesh_name.str().c_str());
+      mesh_ofs.precision(8);
+      pmesh->PrintAsSerial(mesh_ofs);
+   }
+
+   // 11. Store the starting (prior to the optimization) positions.
+   ParGridFunction x0(pfespace);
+   x0 = x;
+
+   // 12. Form the integrator that uses the chosen metric and target.
+   TMOP_QualityMetric *metric = NULL;
+   switch (metric_id)
+   {
+      // T-metrics
+      case 1: metric = new TMOP_Metric_001; break;
```

Comment:
We can leave here only the ones that are used in the sample runs.

Reply:
Good idea. I will leave a shape metric and a shape+size metric for both 2D and 3D.